### ISSUE TITLE
Error with GetCompilationInfoForFile and python 2

### ISSUE THREAD

**Issue by ptrv** [Valloric/ycmd#378]
Hi,

I'm getting the following error when I try to use a `compile_commands.json` and ycmd with python 2:

```python
Traceback (most recent call last):
  File "/home/peter/src/ycmd/third_party/bottle/bottle.py", line 861, in _handle
    return route.call(**args)
  File "/home/peter/src/ycmd/third_party/bottle/bottle.py", line 1734, in wrapper
    rv = callback(*a, **ka)
  File "/home/peter/src/ycmd/ycmd/../ycmd/watchdog_plugin.py", line 105, in wrapper
    return callback( *args, **kwargs )
  File "/home/peter/src/ycmd/ycmd/../ycmd/hmac_plugin.py", line 70, in wrapper
    body = callback( *args, **kwargs )
  File "/home/peter/src/ycmd/ycmd/../ycmd/handlers.py", line 78, in EventNotification
    event_handler )( request_data )
  File "/home/peter/src/ycmd/ycmd/../ycmd/completers/cpp/clang_completer.py", line 318, in OnFileReadyToParse
    flags = self._FlagsForRequest( request_data )
  File "/home/peter/src/ycmd/ycmd/../ycmd/completers/cpp/clang_completer.py", line 381, in _FlagsForRequest
    return self._flags.FlagsForFile( filename, client_data = client_data )
  File "/home/peter/src/ycmd/ycmd/../ycmd/completers/cpp/flags.py", line 95, in FlagsForFile
    client_data )
  File "/home/peter/src/ycmd/ycmd/../ycmd/completers/cpp/flags.py", line 157, in _CallExtraConfFlagsForFile
    return module.FlagsForFile( filename, client_data = client_data )
  File "/home/peter/src/ycmd/cpp/ycm/.ycm_extra_conf.py", line 166, in FlagsForFile
    compilation_info = GetCompilationInfoForFile( filename )
  File "/home/peter/src/ycmd/cpp/ycm/.ycm_extra_conf.py", line 159, in GetCompilationInfoForFile
    return database.GetCompilationInfoForFile( filename )
ArgumentError: Python argument types in
    CompilationDatabase.GetCompilationInfoForFile(CompilationDatabase, unicode)
did not match C++ signature:
    GetCompilationInfoForFile(YouCompleteMe::CompilationDatabase {lvalue}, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
```

**Comment by ptrv** [comment]
Seems that `CompilationDatabase.GetCompilationInfoForFile` does not get along with unicode string. Wrapping the filename in `str()` makes the error disappear.

**Comment by Valloric** [comment]
Yes, that is correct. The C++ layer wants `str` instances on py2 and `bytes` on py3.

**Comment by ptrv** [comment]
Ok, good to know.

The strange thing is that I got this error for the first time since I updated `ycmd` which included #358. The ycm_extra_conf was not new, I used it already for a long time on a project.

**Comment by micbou** [comment]
I can confirm this already happened before #358.

> The ycm_extra_conf was not new, I used it already for a long time on a project. 

Can we see the contents of your `.ycm_extra_conf.py` file?

**Comment by ptrv** [comment]
I first noticed the error after updating `ycmd` with #358 merged

The use the following ycm_extra_conf but after updating ycmd I could reproduce the issue with this projects ycm_extra_conf (modified to use `compile_commands.json`):

```python
import os
import re
import subprocess
import ycm_core

def LoadSystemIncludes():
    regex = re.compile(ur"(?:\#include \<...\> search starts here\:)(?P<list>.*?)(?:End of search list)", re.DOTALL);
    process = subprocess.Popen(['clang', '-v', '-E', '-x', 'c++', '-'], stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE);
    process_out, process_err = process.communicate('');
    output = process_out + process_err;
    includes = [];
    result = re.search(regex, output).group('list').decode('utf-8')
    for p in result.split('\n'):
        p = p.strip();
        if len(p) > 0 and p.find('(framework directory)') < 0:
            includes.append('-isystem');
            includes.append(p);
    return includes;

SOURCE_EXTENSIONS = [ '.cpp', '.cxx', '.cc', '.c', '.m', '.mm' ]
scriptPath = os.path.dirname(os.path.abspath(__file__));
compilation_database_folder = os.path.join(scriptPath, 'Builds/Linux')
database = None if not os.path.exists(compilation_database_folder) else ycm_core.CompilationDatabase(compilation_database_folder)
# database = None
flags = [
]
systemIncludes = LoadSystemIncludes();
flags = flags + systemIncludes;

def MakeRelativePathsInFlagsAbsolute( flags, working_directory ):
  if not working_directory:
    return list(flags)
  new_flags = []
  make_next_absolute = False
  path_flags = ['-isystem', '-I', '-iquote', '--sysroot=']
  for flag in flags:
    new_flag = flag

    if make_next_absolute:
      make_next_absolute = False
      if not flag.startswith('/'):
        new_flag = os.path.join(working_directory, flag)

    for path_flag in path_flags:
      if flag == path_flag:
        make_next_absolute = True
        break

      if flag.startswith(path_flag):
        path = flag[len(path_flag):]
        new_flag = path_flag + os.path.join(working_directory, path)
        break

    if new_flag:
      new_flags.append(new_flag)
  return new_flags

def IsHeaderFile(filename):
  extension = os.path.splitext(filename)[1]
  return extension in ['.h', '.hxx', '.hpp', '.hh']

def GetCompilationInfoForFile(filename):
  # The compilation_commands.json file generated by CMake does not have entries
  # for header files. So we do our best by asking the db for flags for a
  # corresponding source file, if any. If one exists, the flags for that file
  # should be good enough.
  # filename = str(filename)
  if IsHeaderFile(filename):
    basename = os.path.splitext(filename)[0]
    for extension in SOURCE_EXTENSIONS:
      replacement_file = basename + extension
      if not os.path.exists(replacement_file):
        replacement_file = os.path.join(scriptPath, 'Source/Application/Main.cpp')
      compilation_info = database.GetCompilationInfoForFile(replacement_file)
      if compilation_info.compiler_flags_:
        return compilation_info
    return None
  return database.GetCompilationInfoForFile(filename)

def FlagsForFile(filename, **kwargs):
  if database:
    # Bear in mind that compilation_info.compiler_flags_ does NOT return a
    # python list, but a "list-like" StringVec object
    compilation_info = GetCompilationInfoForFile( filename ))
    if not compilation_info:
      return None

    final_flags = MakeRelativePathsInFlagsAbsolute(compilation_info.compiler_flags_, compilation_info.compiler_working_dir_) + systemIncludes
  else:
    final_flags = MakeRelativePathsInFlagsAbsolute(flags, scriptPath)

  return {
    'flags': final_flags,
    'do_cache': True
  }
```

Now I need to wrap the filename with `str()` which is fine

**Comment by micbou** [comment]
Ok, I understand now and it is not specific to your `.ycm_extra_conf.py` file. In my test, I used a unicode string for the compilation database path to raise the error but I didn't think of the `filename` parameter in the `FlagsForFile` function. Since #358, this parameter is now a unicode string. We should probably convert `filename` using `ToCppStringCompatible` before passing it to `FlagsForFile` in [`ycmd/ycmd/completers/cpp/clang_completer.py`](https://github.com/Valloric/ycmd/blob/master/ycmd/completers/cpp/clang_completer.py#L381).

**Comment by Valloric** [comment]
Fix out in #394. It will make it so that you can pass whatever string-like object you want to `CompilationDatabase`.
